p2p/shim.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	return descriptors
+}
+
+// OnStart executes the reactor shim's OnStart hook where we start all the
+// necessary go-routines in order to proxy peer errors and messages.
+func (rs *ReactorShim) OnStart() error {
+	if rs.IsRunning() {
+		rs.proxyPeerErrors()
+		rs.proxyPeerEnvelopes()
+	}
+
+	return nil
+}
+
+// OnStop executes the reactor shim's OnStop hook where all p2p Channels are
+// closed and the PeerUpdateCh is closed.
+func (rs *ReactorShim) OnStop() {
+	for _, c := range rs.Channels {
+		if err := c.Channel.Close(); err != nil {
+			rs.Logger.Error("failed to close channel", "reactor", rs.Name, "ch_id", c.Channel.ID, "err", err)
+		}
+	}
+
+	close(rs.PeerUpdateCh)
+}
+
+// AddPeer sends a PeerUpdate with status PeerStatusUp on the PeerUpdateCh.
+// The embedding reactor must be sure to listen for messages on this channel to
+// handle adding a peer.
+func (rs *ReactorShim) AddPeer(peer Peer) {
+	peerID, err := PeerIDFromString(string(peer.ID()))
+	if err != nil {
+		// It is OK to panic here as we'll be removing the Reactor interface and
+		// Peer type in favor of using a PeerID directly.
+		panic(err)
+	}
+
+	select {
+	case rs.PeerUpdateCh <- PeerUpdate{PeerID: peerID, Status: PeerStatusUp}:
+		rs.Logger.Debug("sent peer update", "reactor", rs.Name, "peer", peerID.String(), "status", PeerStatusUp)
+
+	default:
+		rs.Logger.Debug("dropped peer update", "reactor", rs.Name, "peer", peerID.String(), "status", PeerStatusUp)
+	}
+}
+
+// RemovePeer sends a PeerUpdate with status PeerStatusDown on the PeerUpdateCh.
+// The embedding reactor must be sure to listen for messages on this channel to
+// handle removing a peer.
+func (rs *ReactorShim) RemovePeer(peer Peer, reason interface{}) {
+	peerID, err := PeerIDFromString(string(peer.ID()))
+	if err != nil {
+		// It is OK to panic here as we'll be removing the Reactor interface and
+		// Peer type in favor of using a PeerID directly.
+		panic(err)
+	}
+
+	select {
+	case rs.PeerUpdateCh <- PeerUpdate{PeerID: peerID, Status: PeerStatusDown}:
+		rs.Logger.Debug("sent peer update", "reactor", rs.Name, "peer", peerID.String(), "status", PeerStatusDown)
+
+	default:
+		rs.Logger.Debug("dropped peer update", "reactor", rs.Name, "peer", peerID.String(), "status", PeerStatusDown)
+	}
+}
+
+func (rs *ReactorShim) Receive(chID byte, src Peer, msgBytes []byte) {
+	if !rs.IsRunning() {
+		return
+	}
+
+	cID := ChannelID(chID)
+	channel, ok := rs.Channels[cID]
+	if !ok {
+		rs.Logger.Error("unexpected channel", "peer", src, "ch_id", chID)
+		return
+	}
+
+	msg, err := rs.Codec.Unmarshal(msgBytes)
+	if err != nil {
+		rs.Logger.Error("error decoding message", "peer", src, "ch_id", cID, "msg", msg, "err", err)
+		// TODO: We need a way to handle custom business logic on error.
+		return
+	}
+
+	if err := rs.Codec.Validate(msg); err != nil {
+		rs.Logger.Error("invalid message", "peer", src, "ch_id", cID, "msg", msg, "err", err)
+		// TODO: We need a way to handle custom business logic on error.
```

Comment:
I think this should be the responsibility of the receiver, maybe - often the validation needs access to state (e.g. someone requests a non-existent block). But we could consider adding an interface for the Protobuf message similar to what we do with `Wrapper` which allows for basic self-contained validation.

Reply:
I was thinking of updating the interface as well.

But now that I think of it, I believe having the embedding reactor call `Receive` first and if there is no preliminary error, then call `Receive` on the embedded shim.

Reply:
Not sure I understand this, care to elaborate?

Reply:
```go
func (r *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
	if !r.IsRunning() {
		return
	}

	msg, err := decodeMsg(msgBytes)
	if err != nil {
		r.Logger.Error("error decoding message", "peer", src, "ch_id", chID, "msg", msg, "err", err, "bytes", msgBytes)
		r.Switch.StopPeerForError(src, err)
		return
	}

	if err = validateMsg(msg); err != nil {
		r.Logger.Error("invalid message", "peer", src, "ch_id", chID, "msg", msg, "err", err)
		r.Switch.StopPeerForError(src, err)
		return
	}

	r.shim.Receive(chID, src, msgBytes)
}
```

Reply:
Ah. The state sync reactor shouldn't know about the shim, and neither should the P2P stack. The purpose of a shim is to transparently allow some component to use a new API against an old implementation, and then swap out the old implementation with the new one without changing the component.

`StateSyncReactor` shouldn't implement `Reactor`, and it shouldn't have a `Receive()` method. It'd be used e.g. something like this:

```go
shim := NewReactorShim("StateSync", []ChannelDescriptor{...})
stateSync := NewStateSyncReactor(shim.GetChannel(0x60), shim.GetChannel(0x61), shim.PeerUpdates())
switch.AddReactor(shim)
```

Reply:
Yeah, if we even want to do it.

For message decoding errors, we can just log a message and drop it - or possibly ask the switch to disconnect the peer (I can't remember what reactors do today, we should keep the same behavior). But for semantic errors, it may be better to have the reactor simply validate it itself, so that it can take appropriate action (some errors may need banning, some may need disconnecting, some may need to send a message in response, and others can just be ignored). I suppose we could have the validate function return a peer action for certain errors, but I could see this getting pretty complicated.

Reply:
I have a simple interface for now just to mimic current behavior. We can revisit this later. It probably won't even be a concern because the shim won't exist once this is all over.